pkg/sql/logictest/testdata/logic_test/alter_column_type

```diff
@@ -161,3 +161,86 @@ ALTER TABLE t ALTER s TYPE STRING
 
 statement ok
 DROP TABLE t
+
+# Test cases for ALTER COLUMN TYPE requiring a cast
+
+# Failed alter
+statement ok
+CREATE TABLE t1 (date string)
+
+statement ok
+INSERT INTO t1 VALUES ('hello')
+
+statement error pq: parsing as type timestamp: could not parse "hello"
+ALTER TABLE t1 ALTER COLUMN date TYPE timestamp
+
+subtest ColumnNotPartOfIndex
+
+statement ok
+CREATE TABLE t2 (id int)
+
+statement ok
+INSERT INTO t2 VALUES (1), (2), (3), (4)
+
+statement ok
+ALTER TABLE t2 ALTER COLUMN id TYPE string
+
+query TTBTTTB
+SHOW COLUMNS FROM t2
+----
+id     STRING  true   NULL            ·  {}         false
+rowid  INT8    false  unique_rowid()  ·  {primary}  true
+
+statement ok
+INSERT INTO t2 VALUES ('5')
+
+statement error pq: value type int doesn't match type string of column "id"
+INSERT INTO t2 VALUES (6)
```

Comment:
Missing a test here that the error is what you expect when you insert something that can't be converted back to an int

Reply:
I added that as a test in alter_column_type_test.go, `TestInsertBeforeOldColumnIsDropped`.

My concern for not having it as a logic test is it could possibly be flakey if the garbage collector somehow drops the column quickly - this won't be the case now but possible if the gc changes.